ctapipe/io/serializer.py

```diff
@@ -4,90 +4,255 @@
 
     def __enter__(self):
-        log.debug("Serializing on {0}".format(self.file))
         return self
 
-    def __exit__(self, *args):
-        pass
+    def __exit__(self, exc_type, exc_value, traceback):
+        """
+        Exit the runtime context related to this object.
+        The parameters describe the exception that caused the context to be
+        exited. If the context was exited without an exception,
+        all three arguments will be None.
+        If an exception is supplied, and the method wishes to suppress
+        the exception (i.e., prevent it from being propagated),
+        it should return a true value. Otherwise, the exception will be
+        processed normally upon exit from this method.
+        """
+        self.close()
+
+    def close(self):
+        """
+        Close reader or writer
+        """
+        if self._writer:
+            self._writer.write()
+        elif self._reader:
+            self._reader.close()
 
-    @abstractmethod
     def add_container(self, container):
         """
         Add a container to serializer
+         Raises
+        ------
+        RuntimeError: When Serializer is used as Writer
         """
-        pass
+        if not self._writer:
+            raise RuntimeError('This serializer instance is a reader')
+        self._writer.add_container(container)
+
+    def get_next_container(self):
+        """
+        Returns
+        -------
+        The next container in file
+        Raises
+        ------
+        EOFError:  When end of file is reached without returning Container
+        RuntimeError: When Serializer is used as writer
+        """
+        if not self._reader:
+            raise RuntimeError('This serializer instance is a writer')
+        return self._reader.get_next_container()
 
-    @abstractmethod
     def write(self):
         """
-        Write all containers to outfile
+        Write data to disk OR close it
+        Raises
+        ------
+        RuntimeError: When Serializer is used as reader
+        """
+        if not self._writer:
+            raise RuntimeError('This serializer instance is a reader')
+        self._writer.write()
+
+    def __iter__(self):
+        """
+        Yields
+        ------
+        A container
+        Raises
+        ------
+        RuntimeError: when this Serializer instance is a writer
         """
+        if self._reader:
+            for container in self._reader:
+                yield container
+            raise StopIteration
+        else:
+            raise RuntimeError('This Serializer instance is a writer')
+
+
+class Writer(ABC):
+    def __init__(self, outfile):
+        self.outfile = outfile
+
+    @abstractmethod
+    def add_container(self, container):
+        pass
+
+    @abstractmethod
+    def write(self):
+        pass
+
+
+class Reader(ABC):
+    def __init__(self, infile):
+        self.infile = infile
+
+    @abstractmethod
+    def get_next_container(self):
```

Comment:
python objects can have a __next__ method, which is called for iterators and if you do next(object). I think this would fit quite well here.